### ISSUE TITLE
Spikes; can't nut out the calibration

### ISSUE THREAD

**Issue by garthk** [evanmayer/rtlobs#1]
TL;DR can't figure out how to calibrate; can you translate from radio astronomer terminology to amater-with-an-RTLSDR terminology?

G'day! Tried your 32ea774 with a nooelec [LNA and H1 SAW filter](https://www.nooelec.com/store/sawbird-h1.html) and [SDR](https://www.nooelec.com/store/nesdr-smartee-sdr.html), and I'm getting spikes when I aim my waveguide at the sun or my laptop…

```python
from rtlobs import collect as col
from rtlobs import post_process as post
f, p = col.run_spectrum_int(8192, 2048, 49.6, 2.32e6, 1.420e9, 10)
fig, ax = post.plot_spectrum(f, p, savefig='../images/spectrum_int.png')
```

![spectrum_int_12_at_laptop](https://user-images.githubusercontent.com/15906/75588035-ee3c1f80-5acb-11ea-8ac8-c40d409fb237.png)

… and if I take the LNA off I get another spike bang in the middle, which someone reckons might be some kind of DC spike they've seen in E4000 based RTLSDR though mine has a R820T2 tuner…

![spectrum_int_13_no_lna_sun](https://user-images.githubusercontent.com/15906/75588200-3eb37d00-5acc-11ea-8466-4f01aaf5b2de.png)

… and anyhow it strikes me that subtracting (so to speak) a measurement of the ground or a dummy load might help, but simple subtraction of the numbers on the Y axis might not be numerically valid.

I've followed the link from your `rtlobs/calibrate.py` to [K. O'Neil, Single-Dish Calibration Techniques at Radio Wavelengths](http://articles.adsabs.harvard.edu//full/2002ASPC..278..293O/0000293.000.html) but regret I can't properly follow it. 

Any tips for me and [a few dozen other DIY telescope teams across Australia](http://spaceaustralia.com/news/community-science-project-diy-radio-telescope)? It'd help a great deal if we could make it as simple as, oh, plugging in a 50Ω resistor and running script X to save a profile to a file, and then give that filename to the rest of the code. I'm happy to write arbitrary amounts of Python once I have the vaguest idea what it's supposed to do.

Strikes me the `p_hot` and `p_cold` arguments to `y_factor_cal` come from the second return value of `run_spectrum_int`, but the `t_hot` and `t_cold` arguments require knowledge of our equipment and signal sources that we just don't have. The “system” in the result `t_sys` hints to me that we're summarising the calibration in a way that assumes whatever we're calibrating against is constant regardless of the frequency. Yet, I'm getting that same overall lump shape whether I aim my antenna at the sun or at the ground. 

What's the best way to compensate for our equipments' frequency-specific peculiarities to generate a graph that best represents the signal arriving at the most of the feed horn?

**Comment by evanmayer** [comment]
Thanks for opening an issue! The good news is that I have all of that hardware, so we should be able to work this out together. I'm still figuring this out as well, this project started off in an effort to teach myself this stuff, so I'll give you my best shot, one at a time.

> I'm getting spikes when I aim my waveguide at the sun or my laptop…

First guess is radio frequency interference (RFI) for the spike(s) away from the center frequency you're tuned to. Here's an example. I mirrored your hardware setup: nooelec nesdr smartee, nooelec SAWbird+ H1 Barebones (the one with the switchable 50 Ohm resistor, slightly worse noise figure, otherwise identical). My horn resting with the opening straight down on a concrete floor:
![Figure_2](https://user-images.githubusercontent.com/17206559/75591937-488e9f00-5a3e-11ea-89b4-5eabd9d14160.png)
I see a center spike, and one closer to 1.4205 GHz. Pretty annoying. You'll notice, going back to [this recent article](https://www.rtl-sdr.com/cheap-and-easy-hydrogen-line-radio-astronomy-with-a-rtl-sdr-wifi-parabolic-grid-dish-lna-and-sdrsharp/), that a similar feature appears in their GIF. Their telescope is running the same LNA we are. My guess is that this spike is either very strong RFI, or related to a flaw in the in-band RFI mitigation on the nooelec LNA. Lots of stuff emits RFI in this region of the electromagnetic spectrum, CPUs, GPUs, cell phones, etc. Unfortunately, not much one can do except get far away from the sources. 

Now with my desktop computer turned on, and the horn pointed at the computer:
![Figure_1](https://user-images.githubusercontent.com/17206559/75591967-580de800-5a3e-11ea-992e-f30c0ba81351.png)
Lots of in-band RFI. You've already noticed that they both have the center spike, though. This issue is a bit stickier. Apparently, it's an artifact of some [non-ideal hardware on the SDR](https://dsp.stackexchange.com/questions/40734/what-does-correcting-iq-do). I believe it can be corrected in software. SDR# has a tickbox for it. [rtl-power-fftw](https://github.com/AD-Vega/rtl-power-fftw/blob/master/src/acquisition.cxx) interpolates across the center frequency bin to just ignore the spike (line 376). I haven't implemented either of these yet. If you want to tackle implementation, I'd welcome the pull request!

I'll do separate answers to try to tackle the other problems.

**Comment by evanmayer** [comment]
Translation time: I'll try my best, but I'm also just an amateur with an RTLSDR at this point. Sorry in advance for the wall of text, I guess my lack of brevity comes from how jumbled this is in my head still. It's hard stuff.

Calibration is difficult enough to understand conceptually, and the units people like to work in for this stuff don't help. Small quantities and small differences in power we're dealing with add one complication: working in [decibels (dB)](https://en.wikipedia.org/wiki/Decibel). I still don't have great intuition for working in dB, but at it's core, any power measurement in dB is a relative measurement:

p_db = 10 x log10( p1 / p0 )

with a base-10 log and a factor of 10 out front. Taking the log helps us see, visually, small differences in the spectrum. As far as I understand, the question that calibration is supposed to answer is, if our signal of interest is p1, _what is in the value in the denominator?_ Our telescopes measure power in some arbitrary units, so we have to figure out what that power is, relative to some real-world references.

Another complication is added by plotting the y-axis in units of dB/Hz. This is the [power spectral density](https://en.wikipedia.org/wiki/Spectral_density#Power_spectral_density), a measure of the signal's power content with respect to frequency, normalized by the sample rate. It's a quantity that we like because the y-values shouldn't change if you change the sampling rate of the input signal (the sampling rate of your SDR). This is an excellent article explaining why we'd want to use dB/Hz: [What is a power spectral density (PSD)?](https://community.sw.siemens.com/s/article/what-is-a-power-spectral-density-psd)

So after all that, I'll say that this is good cautious instinct:

> … and anyhow it strikes me that subtracting (so to speak) a measurement of the ground or a dummy load might help, but simple subtraction of the numbers on the Y axis might not be numerically valid.

because [log_n(a) - log_n(b) = log_n(a/b)](https://www.purplemath.com/modules/logrules.htm). 

We're going to have to be careful how we apply a calibration to spectrum. I haven't fully thought through how to accomplish this on an entire spectrum at once (hence, it is on the to-do list). I'd first look at how some of the other packages out there are doing it, and try to sort out if they're valid. [This](http://wvurail.org/dspira/) is probably where I'd start that search. John Makous gave a grcon talk that got [picked up by Hackaday](https://hackaday.com/2019/01/20/radio-telescopes-horn-in-with-gnu-radio/) (slides [here](https://www.gnuradio.org/grcon/grcon18/presentations/open_source_radio_telescopes/2-John_Makous-OSRT.pdf)), and those have a bit of calibration concepts in them. I'll take this as a feature request!

> I've followed the link from your rtlobs/calibrate.py to K. O'Neil, Single-Dish Calibration Techniques at Radio Wavelengths but regret I can't properly follow it.

The crux of that article is "Y-factor calibration". Basically, it's a two-point calibration, where you look at the _total power_* your telescope measures pointing at "hot" thing, and then at a "cold" thing in terms of [the radio emission it has](https://en.wikipedia.org/wiki/Brightness_temperature). Selecting the right hot load and cold load is important, but it is really just f(x) = mx + b, finding a slope. You get a certain power measured looking at a cold load, a greater power measured looking at a hot load, and your calibration draws a line between them, telling you for any power measured, what the brightness temperature of the object is. Typically for amateur H1 stuff, a "cold" thing is the empty night sky, with no Milky Way in sight of your telescope, and a "hot" thing is some nearby grassy ground or trees. Hot and cold loads can also be matched (50 Ohm, usually) resistors hooked up to the input of your receiver: one at room temp (~295 Kelvin), and one in liquid nitrogen (~77 K).

*this means you should use `p_hot = col.run_total_power_int()` and `p_cold = col.run_total_power_int()` to get inputs to `t_sys = cal.y_factor_cal()`. I'll add an example to the .ipynb.

**Comment by evanmayer** [comment]
This kind of discussion probably isn't what GitHub intended the "issues" section for, but I'm fine leaving it open for discussion. If you have a different preferred venue, you can message me on Twitter (search my name) and we can sort it out with you and your group.

**Comment by evanmayer** [comment]
(U) Closing for inactivity.

**Comment by garthk** [comment]
Years later, trying something related, I'm reading that [the “center spike” is a known problem](https://www.rtl-sdr.com/sdrsharp-users-guide/#:~:text=center%20spike) and everyone know knows to correct for DC offset even if they're not sure [what correcting IQ and offsetting DC does](https://dsp.stackexchange.com/a/40740).
